File: code/runtime/hvm/server/loop.rs
// loop.rs - IO action interpreter for server (macOS/Linux/Windows).

use super::bind::{AppInput, HvmApi};
use super::marshal::{from_term, to_term, HvmValue, MarshalContext};
use std::collections::HashMap;

/// Well-known IO constructor name IDs.
pub struct IoNameIds {
  pub io_done: u32,
  pub io_call: u32,
  pub io_bind: u32,
}

pub type NativePrimFn = fn(&MarshalContext, &str, u64) -> HvmValue;

pub struct IoContext<'a> {
  pub marshal: MarshalContext<'a>,
  pub ids: &'a IoNameIds,
  pub prims: &'a HashMap<String, NativePrimFn>,
}

pub fn run_io(ctx: &IoContext, term: u64) -> HvmValue {
  let ids = ctx.ids;

  let mut current = HvmApi::wnf(term);

  loop {
    let name = HvmApi::term_ext(current);

    if name == ids.io_done {
      let loc = HvmApi::term_val(current);
      let value = HvmApi::wnf(HvmApi::heap_read(loc));
      return from_term(&ctx.marshal, value);
    }

    if name == ids.io_call {
      let loc = HvmApi::term_val(current);
      let prim_name_term = HvmApi::wnf(HvmApi::heap_read(loc));
      let prim_name = from_term(&ctx.marshal, prim_name_term);

      let prim_str = match &prim_name {
        HvmValue::Str(s) => s.as_str(),
        _ => panic!("IO.call: expected string prim name"),
      };

      let arg = HvmApi::wnf(HvmApi::heap_read(loc + 1));
      let cont = HvmApi::heap_read(loc + 2);

      let handler = ctx
        .prims
        .get(prim_str)
        .unwrap_or_else(|| panic!("Unknown native primitive: {prim_str}"));

      let result = handler(&ctx.marshal, prim_str, arg);
      let hvm_result = to_term(&ctx.marshal, &result);

      current = HvmApi::wnf(HvmApi::term_new_app(&AppInput {
        f: cont,
        x: hvm_result,
      }));
      continue;
    }

    if name == ids.io_bind {
      let loc = HvmApi::term_val(current);
      let action = HvmApi::heap_read(loc);
      let cont = HvmApi::heap_read(loc + 1);

      let action_result = run_io(ctx, action);
      let hvm_result = to_term(&ctx.marshal, &action_result);

      current = HvmApi::wnf(HvmApi::term_new_app(&AppInput {
        f: cont,
        x: hvm_result,
      }));
      continue;
    }

    panic!("Unknown IO action: {name}");
  }
}

File: code/runtime/hvm/android/loop.kt
// loop.kt - IO action interpreter for Android.

package surf.clue.hvm

typealias NativePrimFn = (MarshalContext, String, Long) -> HvmValue

data class IoNameIds(
    val ioDone: Int,
    val ioCall: Int,
    val ioBind: Int,
)

data class IoContext(
    val marshal: MarshalContext,
    val ids: IoNameIds,
    val prims: Map<String, NativePrimFn>,
)

fun runIo(input: Pair<IoContext, Long>): HvmValue {
    val (ctx, startTerm) = input
    val ids = ctx.ids

    var current = HvmNative.hvmWnf(startTerm)

    while (true) {
        val name = HvmNative.hvmTermExt(current)

        if (name == ids.ioDone) {
            val loc = HvmNative.hvmTermVal(current)
            // Skip magic field at loc+0, value is at loc+1
            val value = HvmNative.hvmWnf(HvmNative.hvmHeapRead(loc + 1))
            return fromTerm(ctx.marshal to value)
        }

        if (name == ids.ioCall) {
            val loc = HvmNative.hvmTermVal(current)
            // Skip magic field at loc+0; func/argm/cont at loc+1/+2/+3
            val primNameTerm = HvmNative.hvmWnf(HvmNative.hvmHeapRead(loc + 1))
            val primName = fromTerm(ctx.marshal to primNameTerm)

            require(primName is HvmValue.Str) {
                "IO.call: expected string prim name"
            }

            val arg = HvmNative.hvmWnf(HvmNative.hvmHeapRead(loc + 2))
            val cont = HvmNative.hvmHeapRead(loc + 3)

            val handler = ctx.prims[primName.value]
                ?: error("Unknown native primitive: ${primName.value}")

            val result = handler(ctx.marshal, primName.value, arg)
            val hvmResult = toTerm(ctx.marshal to result)

            current = HvmNative.hvmWnf(HvmNative.hvmTermNewApp(cont, hvmResult))
            continue
        }

        if (name == ids.ioBind) {
            val loc = HvmNative.hvmTermVal(current)
            // Skip magic field at loc+0; action/cont at loc+1/+2
            val action = HvmNative.hvmHeapRead(loc + 1)
            val cont = HvmNative.hvmHeapRead(loc + 2)

            val actionResult = runIo(ctx to action)
            val hvmResult = toTerm(ctx.marshal to actionResult)

            current = HvmNative.hvmWnf(HvmNative.hvmTermNewApp(cont, hvmResult))
            continue
        }

        error("Unknown IO action: $name")
    }
}
